File: airesources/Kotlin/src/kotlin/halite/Position.kt
package halite

open class Position(val xPos: Double, val yPos: Double) {

    fun getDistanceTo(target: Position): Double {
        val dx = xPos - target.xPos
        val dy = yPos - target.yPos
        return Math.sqrt(Math.pow(dx, 2.0) + Math.pow(dy, 2.0))
    }

    fun orientTowardsInDeg(target: Position): Int {
        return Util.angleRadToDegClipped(orientTowardsInRad(target))
    }

    fun orientTowardsInRad(target: Position): Double {
        val dx = target.xPos - xPos
        val dy = target.yPos - yPos

        return Math.atan2(dy, dx) + 2 * Math.PI
    }

    fun getClosestPoint(target: Entity): Position {
        val radius = target.radius + Constants.MIN_DISTANCE
        val angleRad = target.orientTowardsInRad(this)

        val x = target.xPos + radius * Math.cos(angleRad)
        val y = target.yPos + radius * Math.sin(angleRad)

        return Position(x, y)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val position = other as Position?

        return java.lang.Double.compare(position!!.xPos, xPos) == 0 && java.lang.Double.compare(position.yPos, yPos) == 0
    }

    override fun hashCode(): Int {
        var result: Int
        var temp: Long = java.lang.Double.doubleToLongBits(xPos)
        result = (temp xor temp.ushr(32)).toInt()
        temp = java.lang.Double.doubleToLongBits(yPos)
        result = 31 * result + (temp xor temp.ushr(32)).toInt()

        return result
    }

    override fun toString(): String {
        return "Position($xPos, $yPos)"
    }
}

File: airesources/Dart/lib/src/position.dart
part of hlt;


class Position {
  double _x;
  double _y;

  Position({double x, double y}) : this._x = x, this._y = y;

  double get x => _x;
  double get y => _y;

  bool isAt(Position other) => x == other.x && y == other.y;

  double getDistanceTo(Position target) {
    return sqrt(pow(x - target.x, 2) + pow(y - target.y, 2));
  }

  int orientTowardsInDeg(Position target) {
    return angleRadToDegClipped(orientTowardsInRad(target));
  }

  double orientTowardsInRad(Position target) {
    return atan2(target.y - y, target.x - x) + 2 * PI;
  }

  Position getClosestPoint(final Entity target) {
    final double radius = target.radius + MIN_DISTANCE_FOR_CLOSEST_POINT;
    final double angleRad = target.orientTowardsInRad(this);

    final double x = target.x + radius * cos(angleRad);
    final double y = target.y + radius * sin(angleRad);
    return new Position(x: x, y: y);
  }
}

int toDegrees(double radians) => (radians * 180 / PI).round();

int angleRadToDegClipped(double angleRad) {
  return ((toDegrees(angleRad) % 360) + 360) % 360;
}
